File: app/src/main/java.backup/com/lumiyaviewer/lumiya/render/tex/TextureFormatBridge.java
package com.lumiyaviewer.lumiya.render.tex;

import android.content.Context;
import android.util.Log;

import com.lumiyaviewer.lumiya.render.ModernTextureManager;

import java.io.IOException;
import java.io.InputStream;

/**
 * Integration bridge between legacy JPEG2000 texture system and modern KTX2/Basis Universal system
 * This class provides a migration path and fallback mechanisms
 */
public class TextureFormatBridge {
    private static final String TAG = "TextureFormatBridge";
    
    private static ModernTextureManager modernManager;
    private static boolean modernManagerInitialized = false;
    
    /**
     * Initialize the texture format bridge
     */
    public static void initialize(Context context) {
        try {
            modernManager = new ModernTextureManager(context);
            modernManagerInitialized = true;
            Log.i(TAG, "Modern texture manager initialized successfully");
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize modern texture manager, falling back to legacy system", e);
            modernManagerInitialized = false;
        }
    }
    
    /**
     * Check if modern texture system is available
     */
    public static boolean isModernTextureSystemAvailable() {
        return modernManagerInitialized && modernManager != null;
    }
    
    /**
     * Detect texture format from stream header
     */
    public static TextureFormat detectTextureFormat(InputStream stream) throws IOException {
        // Read first few bytes to detect format
        stream.mark(16);
        byte[] header = new byte[12];
        int bytesRead = stream.read(header);
        stream.reset();
        
        if (bytesRead >= 12) {
            // Check for KTX2 magic bytes: 0xAB, 0x4B, 0x54, 0x58, 0x20, 0x32, 0x30, 0xBB, 0x0D, 0x0A, 0x1A, 0x0A
            if (header[0] == (byte)0xAB && header[1] == 0x4B && header[2] == 0x54 && header[3] == 0x58 &&
                header[4] == 0x20 && header[5] == 0x32 && header[6] == 0x30 && header[7] == (byte)0xBB &&
                header[8] == 0x0D && header[9] == 0x0A && header[10] == 0x1A && header[11] == 0x0A) {
                return TextureFormat.KTX2;
            }
            
            // Check for JPEG2000 magic bytes (JP2 format): 0x00, 0x00, 0x00, 0x0C, 0x6A, 0x50, 0x20, 0x20
            if (header[0] == 0x00 && header[1] == 0x00 && header[2] == 0x00 && header[3] == 0x0C &&
                header[4] == 0x6A && header[5] == 0x50 && header[6] == 0x20 && header[7] == 0x20) {
                return TextureFormat.JPEG2000;
            }
        }
        
        // Default to JPEG2000 for compatibility
        return TextureFormat.JPEG2000;
    }
    
    /**
     * Load texture using appropriate system based on format
     */
    public static TextureData loadTexture(InputStream stream) throws IOException {
        TextureFormat format = detectTextureFormat(stream);
        
        switch (format) {
            case KTX2:
                return loadKTX2Texture(stream);
            case JPEG2000:
                return loadJPEG2000Texture(stream);
            default:
                throw new IOException("Unsupported texture format: " + format);
        }
    }
    
    private static TextureData loadKTX2Texture(InputStream stream) throws IOException {
        if (!isModernTextureSystemAvailable()) {
            throw new IOException("Modern texture system not available");
        }
        
        ModernTextureManager.TextureData modernData = modernManager.loadKTX2Texture(stream);
        
        // Convert to unified TextureData format
        return new TextureData(
            modernData.width,
            modernData.height,
            modernData.data,
            TextureFormat.KTX2,
            modernData.getOpenGLFormat(),
            modernData.isCompressed()
        );
    }
    
    private static TextureData loadJPEG2000Texture(InputStream stream) throws IOException {
        // This would integrate with the existing OpenJPEG system
        // For now, return a placeholder
        throw new IOException("JPEG2000 texture loading not implemented in bridge");
    }
    
    /**
     * Texture format enumeration
     */
    public enum TextureFormat {
        JPEG2000,
        KTX2
    }
    
    /**
     * Unified texture data structure
     */
    public static class TextureData {
        public final int width;
        public final int height;
        public final byte[] data;
        public final TextureFormat format;
        public final int openGLFormat;
        public final boolean compressed;
        
        public TextureData(int width, int height, byte[] data, TextureFormat format, 
                          int openGLFormat, boolean compressed) {
            this.width = width;
            this.height = height;
            this.data = data;
            this.format = format;
            this.openGLFormat = openGLFormat;
            this.compressed = compressed;
        }
        
        @Override
        public String toString() {
            return String.format("TextureData[%dx%d, %s, %s, %d bytes]",
                    width, height, format, compressed ? "compressed" : "uncompressed", data.length);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/tex/TextureFormatBridge.kt
package com.lumiyaviewer.lumiya.render.tex

import android.content.Context
import android.util.Log

import com.lumiyaviewer.lumiya.render.ModernTextureManager

import java.io.IOException
import java.io.InputStream

/**
 * Integration bridge between legacy JPEG2000 texture system and modern KTX2/Basis Universal system
 * This class provides a migration path and fallback mechanisms
 */
class TextureFormatBridge {
    private String TAG = "TextureFormatBridge"
    
    private ModernTextureManager modernManager
    private boolean modernManagerInitialized = false
    
    /**
     * Initialize the texture format bridge
     */
    void initialize(Context context) {
        try {
            modernManager = new ModernTextureManager(context)
            modernManagerInitialized = true
            Log.i(TAG, "Modern texture manager initialized successfully")
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize modern texture manager, falling back to legacy system", e)
            modernManagerInitialized = false
        }
    }
    
    /**
     * Check if modern texture system is available
     */
    boolean isModernTextureSystemAvailable() {
        return modernManagerInitialized && modernManager != null
    }
    
    /**
     * Detect texture format from stream header
     */
    TextureFormat detectTextureFormat(InputStream stream) throws IOException {
        // Read first few bytes to detect format
        stream.mark(16)
        byte[] header = new byte[12]
        int bytesRead = stream.read(header)
        stream.reset()
        
        if (bytesRead >= 12) {
            // Check for KTX2 magic bytes: 0xAB, 0x4B, 0x54, 0x58, 0x20, 0x32, 0x30, 0xBB, 0x0D, 0x0A, 0x1A, 0x0A
            if (header[0] == (byte)0xAB && header[1] == 0x4B && header[2] == 0x54 && header[3] == 0x58 &&
                header[4] == 0x20 && header[5] == 0x32 && header[6] == 0x30 && header[7] == (byte)0xBB &&
                header[8] == 0x0D && header[9] == 0x0A && header[10] == 0x1A && header[11] == 0x0A) {
                return TextureFormat.KTX2
            }
            
            // Check for JPEG2000 magic bytes (JP2 format): 0x00, 0x00, 0x00, 0x0C, 0x6A, 0x50, 0x20, 0x20
            if (header[0] == 0x00 && header[1] == 0x00 && header[2] == 0x00 && header[3] == 0x0C &&
                header[4] == 0x6A && header[5] == 0x50 && header[6] == 0x20 && header[7] == 0x20) {
                return TextureFormat.JPEG2000
            }
        }
        
        // Default to JPEG2000 for compatibility
        return TextureFormat.JPEG2000
    }
    
    /**
     * Load texture using appropriate system based on format
     */
    TextureData loadTexture(InputStream stream) throws IOException {
        TextureFormat format = detectTextureFormat(stream)
        
        switch (format) {
            case KTX2:
                return loadKTX2Texture(stream)
            case JPEG2000:
                return loadJPEG2000Texture(stream)
            default:
                throw new IOException("Unsupported texture format: " + format)
        }
    }
    
    private TextureData loadKTX2Texture(InputStream stream) throws IOException {
        if (!isModernTextureSystemAvailable()) {
            throw new IOException("Modern texture system not available")
        }
        
        ModernTextureManager.TextureData modernData = modernManager.loadKTX2Texture(stream)
        
        // Convert to unified TextureData format
        return new TextureData(
            modernData.width,
            modernData.height,
            modernData.data,
            TextureFormat.KTX2,
            modernData.getOpenGLFormat(),
            modernData.isCompressed()
        )
    }
    
    private TextureData loadJPEG2000Texture(InputStream stream) throws IOException {
        // This would integrate with the existing OpenJPEG system
        // For now, return a placeholder
        throw new IOException("JPEG2000 texture loading not implemented in bridge")
    }
    
    /**
     * Texture format enumeration
     */
    enum class class TextureFormat {
        JPEG2000,
        KTX2
    }
    
    /**
     * Unified texture data structure
     */
    class TextureData {
        int width
        int height
        byte[] data
        TextureFormat format
        int openGLFormat
        boolean compressed
        
        TextureData(int width, int height, byte[] data, TextureFormat format, 
                          int openGLFormat, boolean compressed) {
            this.width = width
            this.height = height
            this.data = data
            this.format = format
            this.openGLFormat = openGLFormat
            this.compressed = compressed
        }
        
        @Override
        String toString() {
            return String.format("TextureData[%dx%d, %s, %s, %d bytes]",
                    width, height, format, compressed ? "compressed" : "uncompressed", data.length)
        }
    }
}
